### PR TITLE
SioConvMultiLRを追加

### PR DESCRIPTION
**Pull request by myxyy**
## 概要

https://github.com/MLShukai/ami/issues/291

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

File: ami/models/components/sioconv_multi_lr.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+class MultiLRMLP(nn.Module):
+    def __init__(
+        self, dim: int, dim_ff_hidden: int, lr_scale: float, weight_decay: float, activation: nn.Module = nn.SiLU()
+    ):
+        super().__init__()
+        self.fc1_weight = nn.Parameter(torch.randn(dim_ff_hidden, dim))
+        self.fc1_bias = nn.Parameter(torch.zeros(dim_ff_hidden))
+        self.fc2_weight = nn.Parameter(torch.randn(dim, dim_ff_hidden))
+        self.fc2_bias = nn.Parameter(torch.zeros(dim))
```
Comment (Geson-anko) [inline ami/models/components/sioconv_multi_lr.py:19]: ここ `nn.Linear`にして、 `linear.apply(init_weights)`でbiasを 0にしたり、`linear.weight.register_hook`で登録できたりしないんですか？

File: ami/models/components/sioconv_multi_lr.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import numpy as np
```
Comment (Geson-anko) [inline ami/models/components/sioconv_multi_lr.py:1]:
```suggestion
import math
```
ただlog演算が欲しいだけなら標準の `math`モジュールを使う方が健全です

File: ami/models/components/sioconv_multi_lr.py [hunk 3]
```diff
@@ -1,50 +1,47 @@
...
         self.fc1_weight = nn.Parameter(torch.randn(dim_ff_hidden, dim))
         self.fc1_bias = nn.Parameter(torch.zeros(dim_ff_hidden))
         self.fc2_weight = nn.Parameter(torch.randn(dim, dim_ff_hidden))
         self.fc2_bias = nn.Parameter(torch.zeros(dim))
         self.activation = activation
 
-        self.lr_scale = torch.exp(torch.linspace(0, np.log(lr_scale), num_head))
-        self.weight_decay = torch.exp(np.log(weight_decay) + torch.linspace(-np.log(lr_scale), 0, num_head))
+        self.lr_scale = torch.exp(torch.linspace(0, np.log(lr_scale), dim_ff_hidden))
+        self.weight_decay = torch.exp(np.log(weight_decay) + torch.linspace(-np.log(lr_scale), 0, dim_ff_hidden))
```
Comment (Geson-anko) [inline ami/models/components/sioconv_multi_lr.py:23]: 質問です。 `lr_scale`と`weight_decay` は`dim_ff_hidden`ごとに設定するんですか？記憶ベクトルは `dim_ff_hidden`個あると思いますが、長期的に更新されるものと、短期的に更新されるものが混在する、という認識であっていますか？

前に 複数のlrを持つMLPが横に並んでいる、という表現をされたと思うのですが、それぞれ固有のLRを持つMLPをNブロック持ち、それを結合する、というものではないのですか？

File: ami/models/components/sioconv_multi_lr.py [hunk 4]
```diff
@@ -1,50 +1,47 @@
...
-            grad.view(dim, num_head, head_dim).mul_(self.lr_scale[None,:,None]) +
-            self.fc2_weight.view(dim, num_head, head_dim).mul_(self.weight_decay[None,:,None]))
-        .view(dim, dim_ff_hidden))
+        self.fc1_weight.register_hook(
+            lambda grad: grad.mul_(self.lr_scale[:, None]) + self.fc1_weight.mul_(self.weight_decay[:, None])
+        )
+        self.fc1_bias.register_hook(lambda grad: grad.mul_(self.lr_scale) + self.fc1_bias.mul_(self.weight_decay))
+        self.fc2_weight.register_hook(
+            lambda grad: grad.mul_(self.lr_scale[None, :]) + self.fc2_weight.mul_(self.weight_decay[None, :])
+        )
```
Comment (Geson-anko) [inline ami/models/components/sioconv_multi_lr.py:31]: `hook`メソッドの引数テンソル自体を変更してしまう `mul_` (アンダースコア付き演算)は非推奨です。
```suggestion
        self.fc1_weight.register_hook(
            lambda grad: grad.mul(self.lr_scale[:, None]) + self.fc1_weight.mul(self.weight_decay[:, None])
        )
        self.fc1_bias.register_hook(lambda grad: grad.mul(self.lr_scale) + self.fc1_bias.mul(self.weight_decay))
        self.fc2_weight.register_hook(
            lambda grad: grad.mul(self.lr_scale[None, :]) + self.fc2_weight.mul(self.weight_decay[None, :])
        )
```

> The hook should not modify its argument, but it can optionally return a new gradient which will be used in place of [grad](https://pytorch.org/docs/stable/generated/torch.Tensor.grad.html#torch.Tensor.grad).

https://pytorch.org/docs/stable/generated/torch.Tensor.register_hook.html

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@myxyy 実装ありがとうございます。 weight decayも必要なんですね。なんか別タスクで精度とか、どれだけ長期的に記憶を持つようになるかとか検証したいですね。

File: ami/models/components/sioconv_multi_lr.py (hunk 4 above)
Comment (myxyy) [reply ami/models/components/sioconv_multi_lr.py]: https://github.com/MLShukai/ami/pull/292/commits/7706641103427c0a13a6b172292dcb7b3bb24045
修正しました

File: ami/models/components/sioconv_multi_lr.py (hunk 1 above)
Comment (myxyy) [reply ami/models/components/sioconv_multi_lr.py]: https://github.com/MLShukai/ami/pull/292/commits/a69a535c339f65d4467566487c690a7609ecd320
`nn.Linear`にしました
`fc2`の`bias`はなんとなくベース学習率で学習するようにしてます

File: ami/models/components/sioconv_multi_lr.py (hunk 3 above)
Comment (myxyy) [reply ami/models/components/sioconv_multi_lr.py:28]: > 記憶ベクトルは`dim_ff_hidden`個あると思いますが、長期的に更新されるものと、短期的に更新されるものが混在する、という認識であっていますか？
はい、その認識であっています。
> それぞれ固有のLRを持つMLPをNブロック持ち、それを結合する、というものではないのですか？
最初は`dim_ff_hidden`をマルチヘッドとして分割してヘッド毎の学習率を変える構造を考えていましたが、ヘッド数のハイパーパラメータが増えることと離散的に学習率が変わることを考えると、隠れ層ニューロン毎に連続的に学習率を変えた方が良いかなと思い現状こうしています

File: ami/models/components/sioconv_multi_lr.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import torch.nn.functional as F
+from torch import Tensor
+
+from .sioconvps import RMSNorm, SioConvPSLayer
+from .stacked_hidden_state import StackedHiddenState
+
+
+class MultiLRMLP(nn.Module):
+    def __init__(
+        self, dim: int, dim_ff_hidden: int, lr_scale: float, weight_decay: float, activation: nn.Module = nn.SiLU()
```
[the remark this answers is not in the document]
Comment (myxyy) [reply ami/models/components/sioconv_multi_lr.py]: https://github.com/MLShukai/ami/pull/292/commits/49f9931bba569680589e13eba5d271539be00bc7
修正しました

File: ami/models/components/sioconv_multi_lr.py (hunk 1 above)
Comment (Geson-anko) [reply ami/models/components/sioconv_multi_lr.py]: まあ fc2のbiasは出力のシフトだけなんで、長期記憶には関わらないと思います。

**Review by Geson-anko — APPROVED** [review]
@myxyy Approveします
